core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -116,22 +117,98 @@ class GroupCoordinator(val brokerId: Int,
-            val group = groupManager.addGroup(new GroupMetadata(groupId, Empty, time))
-            doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
-          }
+      // exist we should reject the request.
+      group = new GroupMetadata(groupId, Empty, time)
+      groupManager.addGroup(group)
+    }
+
+    group.inLock {
+      if (group.currentState == Dead) {
+        // if the group is marked as dead, it means some other thread has just removed the group
+        // from the coordinator metadata; this is likely that the group has migrated to some other
+        // coordinator OR the group is in a transient unstable phase. Let the member retry
+        // joining without the specified member id.
+        responseCallback(joinError(newMemberId, Errors.UNKNOWN_MEMBER_ID))
+        return
+      } else if (protocolNotMatch(group, newMemberId, protocolType, protocols, responseCallback)) {
+        return
+      }
 
-        case Some(group) =>
-          doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
+      // update the newMemberAdded flag to indicate that the join group can be further delayed
+      if (group.is(PreparingRebalance) && group.generationId == 0)
+        group.newMemberAdded = true
+
+      // If member id required, register the member with unknown metadata,
+      // and send back a response to call for another join group request with allocated member id.
+      if (requireKnownMemberId) {
+        group.addUnknownMember(newMemberId, group.groupId, protocolType, protocols)
+        responseCallback(JoinGroupResult(
+          members = Map.empty,
+          memberId = newMemberId,
+          generationId = group.generationId,
+          subProtocol = null,
+          leaderId = group.leaderOrNull,
+          error = Errors.MEMBER_ID_REQUIRED))
+      } else {
+        val member = new MemberMetadata(newMemberId, group.groupId, clientId, clientHost, rebalanceTimeoutMs,
+          sessionTimeoutMs, protocolType, protocols)
+        member.isNew = true
+        group.add(member, responseCallback)
+
+        // The session timeout does not affect new members since they do not have their memberId and
+        // cannot send heartbeats. Furthermore, we cannot detect disconnects because sockets are muted
+        // while the JoinGroup is in purgatory. If the client does disconnect (e.g. because of a request
+        // timeout during a long rebalance), they may simply retry which will lead to a lot of defunct
+        // members in the rebalance. To prevent this going on indefinitely, we timeout JoinGroup requests
+        // for new members. If the new member is still there, we expect it to retry.
+        completeAndScheduleNextExpiration(group, member, NewMemberJoinTimeoutMs)
       }
+      maybePrepareRebalance(group, s"Adding new member $newMemberId")
```

Comment:
In the case that `requireKnownMemberId` is true, I wonder if we should hold off transitioning to `PreparingRebalance` until the member has joined with new memberId?

Reply:
That sounds like the way to go, doesn't it? We can't be sure that a rebalance will happen until we receive the second request with the memberId in that case